File: java/17_threads_and_processes.java
// ============================================================
// Topic:  Multithreading & Multiprocessing  (FOCUS: Thread + ProcessBuilder)
// Summary: Threads with join/synchronization; child processes via ProcessBuilder.
// Run:     java java/17_threads_and_processes.java   (Java 11+ single-file launch)
//
// KEY DIFFERENCES:
//   Python:  threading + multiprocessing / subprocess
//   TS/JS:   Worker threads (node:worker_threads) — no true multiprocessing in browser
//   Rust:    std::thread + std::process::Command
//   Java:    Thread / ExecutorService + ProcessBuilder (no async/await)
//   C++:     std::thread + fork/exec or std::process (C++23)
// ============================================================

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.locks.ReentrantLock;

public class ThreadsAndProcesses {

    // ----- 1. Basic thread — start, run, join -----
    static void basicThread() throws InterruptedException {
        CountDownLatch workerPrinted = new CountDownLatch(1);
        Thread worker = new Thread(() -> {
            System.out.println("worker");                     // Output: worker
            workerPrinted.countDown();
        });
        worker.start();
        workerPrinted.await();
        System.out.println("main");                             // Output: main
        worker.join();
        System.out.println("joined");                           // Output: joined
    }

    // ----- 2. Shared counter — ReentrantLock protects increments -----
    static void lockedCounter() throws InterruptedException {
        final int[] counter = {0};
        ReentrantLock lock = new ReentrantLock();
        Runnable increment = () -> {
            for (int i = 0; i < 1000; i++) {
                lock.lock();
                try {
                    counter[0]++;
                } finally {
                    lock.unlock();
                }
            }
        };
        Thread t1 = new Thread(increment);
        Thread t2 = new Thread(increment);
        t1.start();
        t2.start();
        t1.join();
        t2.join();
        System.out.println("counter: " + counter[0]);         // Output: counter: 2000
    }

    // ----- 3. Two threads — sequential join prints A then B -----
    static void twoThreads() throws InterruptedException {
        Thread threadA = new Thread(() -> System.out.println("A")); // Output: A
        Thread threadB = new Thread(() -> System.out.println("B")); // Output: B
        threadA.start();
        threadA.join();
        threadB.start();
        threadB.join();
    }

    // ----- 4. Child process — ProcessBuilder spawns echo "child" -----
    static void childProcess() throws Exception {
        ProcessBuilder pb;
        if (System.getProperty("os.name").toLowerCase().contains("win")) {
            pb = new ProcessBuilder("cmd", "/c", "echo", "child");
        } else {
            pb = new ProcessBuilder("sh", "-c", "echo child");
        }
        pb.inheritIO();
        Process child = pb.start();
        int exitCode = child.waitFor();
        System.out.println("exit: " + exitCode);                // Output: exit: 0
    }

    public static void main(String[] args) throws Exception {
        // ----- 1. Basic thread -----
        basicThread();

        // ----- 2. Synchronized counter (ReentrantLock) -----
        lockedCounter();

        // ----- 3. Two threads -----
        twoThreads();

        // ----- 4. Child process -----
        childProcess();

        // Cross-language note:
        //   Java has threads (Thread / synchronized / locks) and processes (ProcessBuilder).
        //   No async/await — use CompletableFuture or virtual threads (21+) for async I/O.
        //   ProcessBuilder is like Python subprocess or Rust std::process::Command.
    }
}

File: cpp/17_threads_and_processes.cpp
// ============================================================
// Topic:  Multithreading & Multiprocessing  (FOCUS: std::thread, std::mutex, join, processes)
// Summary: OS threads with mutex sync — separate process via std::system / CreateProcess.
// Run:     g++ -std=c++17 cpp/17_threads_and_processes.cpp -o cpp/17_threads_and_processes && cpp/17_threads_and_processes
//
// KEY DIFFERENCES:
//   C++:     std::thread + std::mutex + join()  |  fork()/exec (Unix) or CreateProcess (Win)
//   Python:  threading.Thread / multiprocessing.Process
//   Java:    Thread / Runnable / ProcessBuilder
//   C#:      Thread / Task.Run / Process.Start
//   Rust:    std::thread::spawn / std::process::Command
//   TS/JS:   Worker threads (worker_threads) / child_process
// ============================================================

#include <atomic>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>  // std::thread, join()

int main(int argc, char* argv[]) {
    // Child process mode — spawned by parent below; only prints "child".
    if (argc > 1 && std::string(argv[1]) == "child") {
        std::cout << "child" << std::endl;
        return 0;
    }

    // ----- 1. std::thread + join() — worker runs, main waits -----
    std::atomic<bool> worker_done{false};
    std::thread worker([&] {
        std::cout << "worker" << std::endl;
        worker_done.store(true);
    });
    while (!worker_done.load()) {
        std::this_thread::yield();  // ensure worker prints before main
    }
    std::cout << "main" << std::endl;
    worker.join();
    std::cout << "joined" << std::endl;
    // Output: worker
    // Output: main
    // Output: joined

    // ----- 2. std::mutex — two threads increment shared counter -----
    int counter = 0;
    std::mutex mtx;
    auto increment = [&](int times) {
        for (int i = 0; i < times; ++i) {
            std::lock_guard<std::mutex> lock(mtx);
            ++counter;
        }
    };
    std::thread inc_a(increment, 1000);
    std::thread inc_b(increment, 1000);
    inc_a.join();
    inc_b.join();
    std::cout << "counter: " << counter << std::endl;
    // Output: counter: 2000

    // ----- 3. Two threads — sequential join keeps A before B -----
    std::thread thread_a([] { std::cout << "A" << std::endl; });
    thread_a.join();
    std::thread thread_b([] { std::cout << "B" << std::endl; });
    thread_b.join();
    // Output: A
    // Output: B

    // ----- 4. Separate process — child prints "child", parent waits -----
#if defined(_WIN32)
    const char* child_cmd = "cpp\\17_threads_and_processes.exe child";
    int exit_code = std::system(child_cmd);
#else
    int exit_code = std::system("cpp/17_threads_and_processes child");
#endif
    std::cout << "exit: " << exit_code << std::endl;
    // Output: child   (printed by child process stdout)
    // Output: exit: 0

    // Cross-language note:
    //   C++17 has no built-in async/await — use std::thread for shared-memory concurrency.
    //   std::mutex + lock_guard protect data races; always join() or detach() every thread.
    //   Processes are separate address spaces — spawn via system(), fork+exec, or CreateProcess.
    //   Unlike Python's GIL, C++ threads can run Python-style CPU work in parallel on multiple cores.
    return 0;
}